feat: multi-expression implemented

guptanshuman124 opened the pull request:
## Description

> Give a brief description of the pull request.

- multi-expression implemented
- panning and zoom implemented
- gridlines and ticks implemented
<img width="1292" height="766" alt="image" src="https://github.com/user-attachments/assets/168ba10a-fd32-45cf-b0b1-bbd336b1c4b3" />


## Semver Changes

- [ ] Patch (bug fix, no new features)
- [ ] Minor (new features, no breaking changes)
- [ ] Major (breaking changes)

## Issues

> List any issues that this pull request closes.
#8 

## Checklist

- [ ] I have read the [Contributing Guidelines](../Contributor_Guide/Contruting.md).

guptanshuman124 committed:
feat: multi-expression implemented

nsh07 commented on src/core/Core/Application.cpp:
```diff
@@ -131,403 +198,203 @@ ExitStatus App::Application::run() {
+            offsetx += dragDelta.x;
+            offsety += dragDelta.y;
+            ImGui::ResetMouseDragDelta(ImGuiMouseButton_Right);
         }
```
Please use the left mouse button instead of right for panning. If that is not possible, do let me know.

nsh07 commented:
Also, implicit functions, polar functions and parametric relations do not work anymore (you have simply removed all the required code for those features). Is this intentional?

guptanshuman124 committed:
fix: panning on left click

guptanshuman124 commented:
@nsh07 I accidently removed the features of implicit function polar function and parametric relation. I would do a commit by tomorrow evening as I was doing some refactoring in code to remove the clutterinside main application.cpp file.

nsh07 commented:
Sure, take your time. Your implementation is really good, and it is exactly what I was looking for.

ArashPartow commented on src/core/Core/expression.hpp:
```diff
@@ -0,0 +1,15 @@
+namespace App::Core {
+
+struct Expression {
+    std::array<char, 1024> expr;  // expression as char array
```
could this instead be a std::string?

guptanshuman124 replied:
I was trying this but faced some issues back then. As I remember that the exprtk compile and expression function require array of character. Not sure about it.

I would try to reuse it and check it out.

ArashPartow replied:
That is not correct.  `std::string` is the primary input parameter for the string representation of an expression for use with the compile method. 

https://www.partow.net/programming/exprtk/exprtk.hpp_.html#line_24640

guptanshuman124 replied:
Ok got it, will change the implementation. There might be something else back when I was structuring it. Thanks for the help

ArashPartow commented on src/core/Core/Application.cpp:
```diff
@@ -131,403 +198,203 @@ ExitStatus App::Application::run() {
+        for (size_t i = 0; i < functions.size(); ++i) {
+            if (!functions[i].visible) continue;
+            expressions[i].register_symbol_table(symbolTable);
+            parser.compile(functions[i].expr.data(), expressions[i]);
```
Make sure to check the return value of the compile method. if it is false the function should not be further evaluated.

ArashPartow replied:
Functions should only be compiled when they change. The Expression struct should perhaps hold the exprtk::expression instance.  As constructing/destructing exprtk::parser/symbol_table/expression types on a per frame basis will be very expensive.

nsh07 commented:
Thank you so much for your review @ArashPartow

@guptanshuman124 please also resolve these change requests while you're working on the PR

guptanshuman124 commented:
> Thank you so much for your review @ArashPartow
> 
> @guptanshuman124 please also resolve these change requests while you're working on the PR

@ArashPartow thank you for the valuable feedback.I will take care of it.

@nsh07 my current progress 

- made the refactoring of code for removing the cluterness(adding new plotter and expression files for structuring)
- reimplemented polar,parametric and explicit plotting
- implicit functions and inequality are left willl try to do as soon as possible
- added a remove function button 
- added random color generation on adding a new function
<img width="1293" height="759" alt="image" src="https://github.com/user-attachments/assets/72af609a-7221-4eee-8da6-da5bad078997" />
<img width="241" height="329" alt="image" src="https://github.com/user-attachments/assets/f82fb09f-598f-4875-aca5-4b2f7ad1e01f" />
